File: modules/adapters/src/main/java/com/streamfirst/iceberg/hybrid/adapters/InMemorySyncAdapter.java
package com.streamfirst.iceberg.hybrid.adapters;

import com.streamfirst.iceberg.hybrid.domain.*;
import com.streamfirst.iceberg.hybrid.ports.SyncPort;
import lombok.extern.slf4j.Slf4j;

import java.time.Instant;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;
import java.util.stream.Collectors;

/**
 * In-memory implementation of SyncPort for testing and development.
 * Provides simple event queuing and status tracking using in-memory collections.
 * Data is lost when the application stops - not suitable for production use.
 */
@Slf4j
public class InMemorySyncAdapter implements SyncPort {
    
    private final Map<String, SyncEvent> events = new ConcurrentHashMap<>();
    private final Map<Region, List<String>> pendingEventsByRegion = new ConcurrentHashMap<>();
    private final Map<TableId, List<String>> eventsByTable = new ConcurrentHashMap<>();
    private final AtomicLong eventCounter = new AtomicLong(1);

    @Override
    public void publishSyncEvent(SyncEvent event) {
        log.debug("Publishing sync event: {}", event);
        
        // Store the event
        events.put(event.getEventId(), event);
        
        // Add to pending queue for target region
        pendingEventsByRegion.computeIfAbsent(event.getTargetRegion(), k -> new ArrayList<>())
                            .add(event.getEventId());
        
        // Add to table history
        eventsByTable.computeIfAbsent(event.getTableId(), k -> new ArrayList<>())
                    .add(event.getEventId());
        
        log.info("Published sync event {} for table {} to region {}", 
                 event.getEventId(), event.getTableId(), event.getTargetRegion());
    }

    @Override
    public List<SyncEvent> getPendingEvents(Region targetRegion) {
        List<String> eventIds = pendingEventsByRegion.get(targetRegion);
        if (eventIds == null) {
            log.debug("No pending events found for region {}", targetRegion);
            return List.of();
        }
        
        List<SyncEvent> pendingEvents = eventIds.stream()
                .map(events::get)
                .filter(Objects::nonNull)
                .filter(event -> event.getStatus() == SyncEvent.Status.PENDING)
                .sorted(Comparator.comparing(SyncEvent::getCreatedAt))
                .toList();
        
        log.debug("Found {} pending events for region {}", pendingEvents.size(), targetRegion);
        return pendingEvents;
    }

    @Override
    public void updateEventStatus(String eventId, SyncEvent.Status status) {
        SyncEvent event = events.get(eventId);
        if (event == null) {
            throw new IllegalArgumentException("Event not found: " + eventId);
        }
        
        log.debug("Updating event {} status from {} to {}", 
                 eventId, event.getStatus(), status);
        
        SyncEvent updatedEvent = event.withStatus(status, Instant.now());
        events.put(eventId, updatedEvent);
        
        // Remove from pending queue if completed or failed
        if (status == SyncEvent.Status.COMPLETED || status == SyncEvent.Status.FAILED) {
            List<String> pendingIds = pendingEventsByRegion.get(event.getTargetRegion());
            if (pendingIds != null) {
                pendingIds.remove(eventId);
            }
        }
        
        log.debug("Updated event {} status to {}", eventId, status);
    }

    @Override
    public SyncEvent createMetadataSyncEvent(TableMetadata metadata, Region targetRegion) {
        String eventId = "metadata-sync-" + eventCounter.getAndIncrement();
        Instant now = Instant.now();
        
        SyncEvent event = new SyncEvent(
            eventId,
            SyncEvent.Type.METADATA_SYNC,
            metadata.getTableId(),
            metadata.getCommitId(),
            metadata.getSourceRegion(),
            targetRegion,
            SyncEvent.Status.PENDING,
            now,
            now
        );
        
        log.debug("Created metadata sync event {} for table {} from {} to {}", 
                 eventId, metadata.getTableId(), metadata.getSourceRegion(), targetRegion);
        return event;
    }

    @Override
    public SyncEvent createDataSyncEvent(TableMetadata metadata, List<String> dataFiles, Region targetRegion) {
        String eventId = "data-sync-" + eventCounter.getAndIncrement();
        Instant now = Instant.now();
        
        SyncEvent event = new SyncEvent(
            eventId,
            SyncEvent.Type.DATA_SYNC,
            metadata.getTableId(),
            metadata.getCommitId(),
            metadata.getSourceRegion(),
            targetRegion,
            SyncEvent.Status.PENDING,
            now,
            now
        );
        
        log.debug("Created data sync event {} for table {} with {} files from {} to {}", 
                 eventId, metadata.getTableId(), dataFiles.size(), 
                 metadata.getSourceRegion(), targetRegion);
        return event;
    }

    @Override
    public List<SyncEvent> getEventHistory(TableId tableId, Region region) {
        List<String> eventIds = eventsByTable.get(tableId);
        if (eventIds == null) {
            log.debug("No events found for table {}", tableId);
            return List.of();
        }
        
        List<SyncEvent> tableEvents = eventIds.stream()
                .map(events::get)
                .filter(Objects::nonNull)
                .filter(event -> event.getTargetRegion().equals(region))
                .sorted(Comparator.comparing(SyncEvent::getCreatedAt))
                .toList();
        
        log.debug("Found {} events for table {} in region {}", 
                 tableEvents.size(), tableId, region);
        return tableEvents;
    }

    @Override
    public List<SyncEvent> getFailedEvents(Region region) {
        List<String> eventIds = pendingEventsByRegion.get(region);
        if (eventIds == null) {
            log.debug("No events found for region {}", region);
            return List.of();
        }
        
        List<SyncEvent> failedEvents = eventIds.stream()
                .map(events::get)
                .filter(Objects::nonNull)
                .filter(event -> event.getStatus() == SyncEvent.Status.FAILED)
                .sorted(Comparator.comparing(SyncEvent::getUpdatedAt).reversed())
                .toList();
        
        log.debug("Found {} failed events for region {}", failedEvents.size(), region);
        return failedEvents;
    }

    @Override
    public void retryFailedEvent(String eventId) {
        SyncEvent event = events.get(eventId);
        if (event == null) {
            throw new IllegalArgumentException("Event not found: " + eventId);
        }
        
        if (event.getStatus() != SyncEvent.Status.FAILED) {
            throw new IllegalArgumentException("Event " + eventId + " is not in FAILED status");
        }
        
        log.info("Retrying failed event {}", eventId);
        
        // Reset status to PENDING
        SyncEvent retriedEvent = event.withStatus(SyncEvent.Status.PENDING, Instant.now());
        events.put(eventId, retriedEvent);
        
        // Add back to pending queue if not already there
        List<String> pendingIds = pendingEventsByRegion.computeIfAbsent(
            event.getTargetRegion(), k -> new ArrayList<>());
        if (!pendingIds.contains(eventId)) {
            pendingIds.add(eventId);
        }
        
        log.info("Retried event {} - reset to PENDING status", eventId);
    }

    /**
     * Clears all sync data. Useful for testing.
     */
    public void clear() {
        log.info("Clearing all sync data");
        events.clear();
        pendingEventsByRegion.clear();
        eventsByTable.clear();
        eventCounter.set(1);
    }

    /**
     * Gets the total number of events across all states.
     */
    public int getTotalEventCount() {
        return events.size();
    }

    /**
     * Gets event count by status for monitoring.
     */
    public Map<SyncEvent.Status, Long> getEventCountByStatus() {
        return events.values().stream()
                .collect(Collectors.groupingBy(
                    SyncEvent::getStatus, 
                    Collectors.counting()));
    }

    /**
     * Gets event count by type for monitoring.
     */
    public Map<SyncEvent.Type, Long> getEventCountByType() {
        return events.values().stream()
                .collect(Collectors.groupingBy(
                    SyncEvent::getType, 
                    Collectors.counting()));
    }

    /**
     * Gets all events for debugging purposes.
     */
    public Collection<SyncEvent> getAllEvents() {
        return new ArrayList<>(events.values());
    }
}

File: modules/adapters/src/main/scala/com/streamfirst/iceberg/hybrid/adapters/InMemorySyncAdapter.scala
package com.streamfirst.iceberg.hybrid.adapters

import com.streamfirst.iceberg.hybrid.domain.DomainError.SyncError
import com.streamfirst.iceberg.hybrid.domain.{
  DomainError,
  EventId,
  PaginatedResult,
  PaginationRequest,
  Region,
  StoragePath,
  SyncEvent,
  TableMetadata
}
import com.streamfirst.iceberg.hybrid.ports.SyncPort
import zio.stream.ZStream
import zio.{IO, Ref, ZIO, ZLayer}

import java.time.Instant

/** In-memory implementation of SyncPort for testing and development. NOT suitable for production
  * use - data is lost on restart. Uses ZIO Ref for thread-safe atomic operations.
  */
final case class InMemorySyncAdapter(events: Ref[Map[EventId, SyncEvent]]) extends SyncPort:

  override def publishSyncEvent(event: SyncEvent): IO[SyncError, Unit] =
    events.update(_ + (event.eventId -> event)) *>
      ZIO.logDebug(s"Published sync event ${event.eventId}")

  override def getSyncEvents(predicate: SyncEvent => Boolean): IO[SyncError, List[SyncEvent]] =
    events.get.map(_.values.filter(predicate).toList.sortBy(_.createdAt))

  override def updateEventStatus(eventId: EventId, status: SyncEvent.Status): IO[SyncError, Unit] =
    for {
      now <- ZIO.succeed(Instant.now())
      result <- events.modify { m =>
        m.get(eventId) match
          case Some(ev) =>
            val updated = ev.withStatus(status, now)
            (Right(()): Either[SyncError, Unit], m.updated(eventId, updated))
          case None => (Left(DomainError.SyncEventNotFound(eventId)), m)
      }
      _ <- ZIO.fromEither(result)
      _ <- ZIO.logDebug(s"Updated event $eventId status to $status")
    } yield ()

  override def createMetadataSyncEvent(
      metadata: TableMetadata,
      targetRegion: Region
  ): IO[SyncError, SyncEvent] =
    for {
      event <- ZIO.succeed(SyncEvent.create(
        SyncEvent.Type.MetadataSync,
        metadata.tableId,
        metadata.commitId,
        metadata.sourceRegion,
        targetRegion
      ))
      _ <- events.update(_.updated(event.eventId, event))
      _ <- ZIO.logDebug(s"Created metadata sync event ${event.eventId}")
    } yield event

  override def createDataSyncEvent(
      metadata: TableMetadata,
      dataFiles: List[StoragePath],
      targetRegion: Region
  ): IO[SyncError, SyncEvent] =
    for {
      event <- ZIO.succeed(SyncEvent.create(
        SyncEvent.Type.DataSync,
        metadata.tableId,
        metadata.commitId,
        metadata.sourceRegion,
        targetRegion
      ))
      _ <- events.update(_.updated(event.eventId, event))
      _ <- ZIO.logDebug(s"Created data sync event ${event.eventId}")
    } yield event

  override def retryFailedEvent(eventId: EventId): IO[SyncError, Unit] =
    for {
      now <- ZIO.succeed(Instant.now())
      result <- events.modify { m =>
        m.get(eventId) match
          case Some(ev) if ev.status == SyncEvent.Status.Failed =>
            val retried = ev.withStatus(SyncEvent.Status.Pending, now)
            (Right(()): Either[SyncError, Unit], m.updated(eventId, retried))
          case Some(ev) =>
            // keep map unchanged
            (Left(DomainError.SyncEventNotFound(eventId)): Either[SyncError, Unit], m)
          case None => (Left(DomainError.SyncEventNotFound(eventId)): Either[SyncError, Unit], m)
      }
      _ <- ZIO.fromEither(result)
      _ <- ZIO.logDebug(s"Retried failed event $eventId")
    } yield ()

  override def getSyncEventsStream(predicate: SyncEvent => Boolean): ZStream[Any, SyncError, SyncEvent] =
    ZStream.fromIterableZIO(
      events.get.map(_.values.filter(predicate).toList.sortBy(_.createdAt))
    )

  override def getSyncEventsPaginated(
      predicate: SyncEvent => Boolean,
      pagination: PaginationRequest
  ): IO[SyncError, PaginatedResult[SyncEvent]] =
    events.get.map { eventsMap =>
      val filteredEvents = eventsMap.values.filter(predicate).toList.sortBy(_.createdAt)
      val startIndex = pagination.continuationToken.flatMap(_.toIntOption).getOrElse(0)
      val endIndex = Math.min(startIndex + pagination.pageSize, filteredEvents.size)
      val pageEvents = filteredEvents.slice(startIndex, endIndex)
      val nextToken = if endIndex < filteredEvents.size then Some(endIndex.toString) else None
      
      PaginatedResult(
        items = pageEvents,
        continuationToken = nextToken,
        hasMore = endIndex < filteredEvents.size
      )
    }

object InMemorySyncAdapter:
  val live: ZLayer[Any, Nothing, SyncPort] = ZLayer.fromZIO {
    for { ref <- Ref.make(Map.empty[EventId, SyncEvent]) } yield InMemorySyncAdapter(ref)
  }
